Finetune sentence transformers with Modal

justinliang1020 opened the pull request:
initial finetune sentence transformers with modal

PR details in README.md

See here for explanations on choice of dataset and loss function: https://www.sbert.net/examples/training/quora_duplicate_questions/README.html

justinliang1020 committed:
initial finetune sentence transformers

justinliang1020 commented on applications/sentence-transformers-ft/main.py:
```diff
@@ -0,0 +1,137 @@
+
+    test_examples = get_quora_examples("test")
+
+    evaluator = evaluation.BinaryClassificationEvaluator.from_input_examples(
```
Might change this to `InformationRetrievalEvaluator` since it's more similar to the MNR loss being used here.
https://www.sbert.net/docs/package_reference/evaluation.html#sentence_transformers.evaluation.InformationRetrievalEvaluator

source:
"MultipleNegativesRankingLoss is especially suitable for Information Retrieval / Semantic Search."
https://www.sbert.net/examples/training/quora_duplicate_questions/README.html#multiplenegativesrankingloss

justinliang1020 committed:
use full dataset

justinliang1020 committed:
save evals

justinliang1020 committed:
refactor dataset code

justinliang1020 committed:
real finetune params

justinliang1020 commented:
Still editing this so no need to review too deeply yet

justinliang1020 committed:
update readme with real finetune

justinliang1020 committed:
cleanup

jxnl commented on applications/sentence-transformers-ft/main.py:
```diff
@@ -0,0 +1,97 @@
+    model = SentenceTransformer(model_id)
+
+    train_examples = []
+    for i in range(train_dataset.num_rows):
```
use for item in list

rather than 

```python
for i in range(len(list)):
    list[i]
```

this could mostly be done with a list comprehension like 


```
test_examples = [InputExample(texts=[item["questions"]["text"][0], item["questions"]["text"][1]], label=int(item["is_duplicate"])) for item in test_dataset]
evaluator = evaluation.BinaryClassificationEvaluator.from_input_examples(test_examples)
```

justinliang1020 replied:
I coded it that way because I was temporarily training on a fraction of the dataset ( I.e train_dataset.num_rows // 100), but yea I'll change it to be more pythonic

jxnl reviewed (commented):
where did optuna code go?
also, you should add another Dense layer so we can at least grid search the final dimensions like his:
https://www.sbert.net/docs/training/overview.html#creating-networks-from-scratch

(not sure if pooling is needed depending on model

```python
word_embedding_model = models.Transformer(...)
dense_model = models.Dense(in_features=word_embedding_model.get_sentence_embedding_dimension(), out_features=???, activation_function=???)

model = SentenceTransformer(modules=[word_embedding_model, dense_model])
```

justinliang1020 commented:
> where did optuna code go?
> 
> also, you should add another Dense layer so we can at least grid search the final dimensions like his:
> 
> https://www.sbert.net/docs/training/overview.html#creating-networks-from-scratch
> 
> 
> 
> (not sure if pooling is needed depending on model
> 
> 
> 
> ```python
> 
> word_embedding_model = models.Transformer(...)
> 
> dense_model = models.Dense(in_features=word_embedding_model.get_sentence_embedding_dimension(), out_features=???, activation_function=???)
> 
> 
> 
> model = SentenceTransformer(modules=[word_embedding_model, dense_model])
> 
> ```

I'm adding back the optuna code rn because I removed it in the sentence transformers refactor. Can add the dense layer too 👍

justinliang1020 commented:
Preliminarily testing out adding a dense linear layer after the embedding model (same hyperparameters, just modifying the dense layer).

No dense layer(base embedding size= 384): 0.8644
Dense Layer(size=128): 0.8626
Dense Layer(size=600): 0.8635

Hyperparameters
base model: bge-small-en-v1.5
epochs: 10
batch_size: 64
dataset_fraction: 10 (only used 1/10 of dataset)
others: default

Will do some more rigorous tests with optuna

justinliang1020 committed:
add dense layer

justinliang1020 committed:
small refactor

justinliang1020 committed:
finetune refactor

justinliang1020 committed:
big refactor

justinliang1020 commented:
did a refactor to make the finetune code re-usable for both regular modal finetune runs and modal optuna finetune runs. will add optuna next

jxnl commented on applications/sentence-transformers-ft/finetune.py:
```diff
@@ -0,0 +1,103 @@
+    train_dataset = train_test_split["train"]
+    test_dataset = train_test_split["test"]
+
+    if use_dense_layer:
```
prosb do something liek 

if dense_out_features:
   ...
   
   saves a variable

justinliang1020 replied:
removed the `use_dense_layer` variable

jxnl commented on applications/sentence-transformers-ft/finetune.py:
```diff
@@ -0,0 +1,103 @@
+        dense_model = models.Dense(
+            in_features=embedding_model.get_sentence_embedding_dimension(),
+            out_features=dense_out_features,
+            activation_function=nn.Tanh(),
```
we can probs search this too, 

tanh, sigmoid, relu, etc.

justinliang1020 replied:
Made `activation_function` a function parameter for finetuning so it can be added to optuna searc

jxnl commented on applications/sentence-transformers-ft/finetune.py:
```diff
@@ -0,0 +1,103 @@
+    )
+
+    # evaluator.name is used for how the file name is saved
+    evaluator.csv_file = "binary_classification_evaluation_pre_train" + "_results.csv"
```
should we be making edits to the attribute?

justinliang1020 replied:
I think it's ok to modify this attribute. the `self.csv_file` variable in the `BinaryClassificationEvaluator` class is only used once when naming the csv file that the evaluator writes results to https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/evaluation/BinaryClassificationEvaluator.py#L97
```
        if output_path is not None and self.write_csv:
            csv_path = os.path.join(output_path, self.csv_file)
            if not os.path.isfile(csv_path):
                with open(csv_path, newline='', mode="w", encoding="utf-8") as f:
                    writer = csv.writer(f)
                    writer.writerow(self.csv_headers)
                    writer.writerow(file_output_data)
            else:
                with open(csv_path, newline='', mode="a", encoding="utf-8") as f:
                    writer = csv.writer(f)
                    writer.writerow(file_output_data)
```

In this case we edit it to manually save the pre-training and post-training evaluations into different CSV files so we can see both of them and compare. Also, after modifying the pre-training csv_file name, I revert it to it's default name (note we don't set a name attribute so that part is empty)
https://github.com/UKPLab/sentence-transformers/blob/master/sentence_transformers/evaluation/BinaryClassificationEvaluator.py#L51
```self.csv_file = "binary_classification_evaluation" + ("_"+name if name else '') + "_results.csv"``` 

( I do think the way this code is structured is a little ugly, I'll look into revising it)

jxnl commented on applications/sentence-transformers-ft/modal.py:
```diff
@@ -0,0 +1,60 @@
+# Functions for Modal Image build steps (cache the model and dataset)
+# NOTE: this can be removed for simplicity, this is currently here for faster devloop times when repeatedly running the function
+def download_model():
+    model_id = "BAAI/bge-small-en-v1.5"
```
should be a script variable

justinliang1020 replied:
Changed to make it a script variable https://github.com/567-labs/fastllm/pull/28/files#diff-7a2ff54b496ec93dabf2bf46fff30a5afbecca3544de44e4f30dc21ed4c84459R12

jxnl commented on applications/sentence-transformers-ft/modal.py:
```diff
@@ -0,0 +1,60 @@
+    _allow_background_volume_commits=True,
+)
+def finetune_modal():
+    model_id = "BAAI/bge-small-en-v1.5"
```
if i want to change the model i currently have to change 3 variables and make sure they line up

justinliang1020 replied:
changed it to just have `MODEL_ID` var at the top of the script
https://github.com/567-labs/fastllm/pull/28/files#diff-7a2ff54b496ec93dabf2bf46fff30a5afbecca3544de44e4f30dc21ed4c84459R7

justinliang1020 committed:
optional dense out features

justinliang1020 committed:
parameterize activation function

justinliang1020 committed:
cleanup finetune

justinliang1020 committed:
rename modal-main and change caching

justinliang1020 committed:
initial optuna

justinliang1020 committed:
add scheduler param

justinliang1020 commented:
Added initial optuna code, currently running a parallel training job

Thinking about adding tensorboard next to visualize training better, also gonna try to addd some more hyperparaeters to tune

EDIT: logs from training run (3 GPU containers running 4 trials each)
```
[I 2023-12-20 18:32:13,786] trial 3 finished with value: 0.3887035871230916 and parameters: {'dense_out_features': 215, 'activation': 'Sigmoid', 'epochs': 7, 'scheduler': 'warmupconstant'}. Best is trial 3 with value: 0.3887035871230916.
Expand
[I 2023-12-20 18:45:35,210] trial 1 finished with value: 0.9038039217033202 and parameters: {'dense_out_features': 200, 'activation': 'Tanh', 'epochs': 8, 'scheduler': 'warmupconstant'}. Best is trial 1 with value: 0.9038039217033202.
[I 2023-12-20 18:45:41,648] trial 2 finished with value: 0.5007514251529964 and parameters: {'dense_out_features': 240, 'activation': 'Sigmoid', 'epochs': 8, 'scheduler': 'warmuplinear'}. Best is trial 1 with value: 0.9038039217033202.
[I 2023-12-20 19:08:50,587] trial 0 finished with value: 0.9059162867150665 and parameters: {'dense_out_features': 547, 'activation': 'Tanh', 'epochs': 12, 'scheduler': 'warmupcosine'}. Best is trial 0 with value: 0.9059162867150665.
[I 2023-12-20 19:27:48,786] trial 4 finished with value: 0.9049505090539349 and parameters: {'dense_out_features': 261, 'activation': 'Tanh', 'epochs': 11, 'scheduler': 'warmupcosine'}. Best is trial 0 with value: 0.9059162867150665.
[I 2023-12-20 19:28:52,762] trial 6 finished with value: 0.9022449082914484 and parameters: {'dense_out_features': 488, 'activation': 'Tanh', 'epochs': 7, 'scheduler': 'warmupconstant'}. Best is trial 0 with value: 0.9059162867150665.
[I 2023-12-20 19:41:59,999] trial 5 finished with value: 0.5521938547442502 and parameters: {'dense_out_features': 420, 'activation': 'Sigmoid', 'epochs': 9, 'scheduler': 'warmupconstant'}. Best is trial 0 with value: 0.9059162867150665.
[I 2023-12-20 20:03:55,789] trial 7 finished with value: 0.9042513916653487 and parameters: {'dense_out_features': 624, 'activation': 'ReLU', 'epochs': 9, 'scheduler': 'warmuplinear'}. Best is trial 0 with value: 0.9059162867150665.
[W 2023-12-20 20:09:51,500] trial 8 failed with parameters: {'dense_out_features': 512, 'activation': 'Sigmoid', 'epochs': 10, 'scheduler': 'warmuplinear'} because of the following error: KeyboardInterrupt().
Traceback (most recent call last):
  File "/usr/local/lib/python3.11/site-packages/optuna/study/_optimize.py", line 200, in _run_trial
    value_or_values = func(trial)
                      ^^^^^^^^^^^
  File "/root/modal-optuna.py", line 109, in <lambda>
    study.optimize(lambda trial: objective(trial), n_trials=N_trialS)
                                 ^^^^^^^^^^^^^^^^
  File "/root/modal-optuna.py", line 82, in objective
    res = finetune(
          ^^^^^^^^^
  File "/root/finetune.py", line 89, in finetune
    model.fit(
  File "/usr/local/lib/python3.11/site-packages/sentence_transformers/SentenceTransformer.py", line 721, in fit
    loss_value = loss_model(features, labels)
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1518, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/nn/modules/module.py", line 1527, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/sentence_transformers/losses/OnlineContrastiveLoss.py", line 49, in forward
    negative_pairs = negs[negs < (poss.max() if len(poss) > 1 else negs.mean())]
                                                ^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/torch/_tensor.py", line 964, in __len__
    def __len__(self):
    
KeyboardInterrupt
[W 2023-12-20 20:09:51,511] trial 8 failed with value None.
[I 2023-12-20 20:36:01,353] trial 9 finished with value: 0.9072778026189381 and parameters: {'dense_out_features': 228, 'activation': 'Tanh', 'epochs': 11, 'scheduler': 'warmupconstant'}. Best is trial 9 with value: 0.9072778026189381.
[I 2023-12-20 20:38:17,410] trial 10 finished with value: 0.42512456916972935 and parameters: {'dense_out_features': 361, 'activation': 'Sigmoid', 'epochs': 9, 'scheduler': 'warmuplinear'}. Best is trial 9 with value: 0.9072778026189381.
[I 2023-12-20 20:46:55,799] trial 11 finished with value: 0.9018604948806459 and parameters: {'dense_out_features': 260, 'activation': 'Tanh', 'epochs': 7, 'scheduler': 'warmupconstant'}. Best is trial 9 with value: 0.9072778026189381.
[I 2023-12-20 20:53:35,418] trial 12 finished with value: 0.8996697522769658 and parameters: {'dense_out_features': 470, 'activation': 'ReLU', 'epochs': 7, 'scheduler': 'warmupcosine'}. Best is trial 9 with value: 0.9072778026189381.
[I 2023-12-20 22:07:26,771] trial 13 finished with value: 0.9048375670467318 and parameters: {'dense_out_features': 787, 'activation': 'Tanh', 'epochs': 12, 'scheduler': 'warmupcosine'}. Best is trial 9 with value: 0.9072778026189381.
[I 2023-12-20 23:15:17,333] trial 14 finished with value: 0.9052930537882344 and parameters: {'dense_out_features': 344, 'activation': 'Tanh', 'epochs': 11, 'scheduler': 'warmupcosine'}. Best is trial 9 with value: 0.9072778026189381.
```

jxnl commented:
> Added initial optuna code, currently running a parallel training job
> 
> Thinking about adding tensorboard next to visualize training better, also gonna try to addd some more hyperparaeters to tune

try to use wandb instead, a lot easier in some ways

justinliang1020 commented:
Main take-away for me right now after doing some hyperparameter testing is that the you can have a pretty low dense layer output # (like 1/2 the size of the base embedding but will test more) and still get good performance. Not sure if this is specific to this embedding model/dataset/testing methodology but is definitely interesting if you can halve your embedding size and still get good retrieval (lots of space saved in your vectordb).

Also, having a parameter count higher than the base model doesn't seem to improve performance very much (?)

logs here: https://github.com/567-labs/fastllm/pull/28#issuecomment-1864984328

justinliang1020 commented:
Another hyperparameter we could test is the base model itself. (another bge variant or another model entirely)


Also think that doing parallel evaluations with different base models on the user's custom data is very cool. An open problem in the RAG space has been "what embedding model should I use?" and this seems like an elegant way to approach solving it.

justinliang1020 commented on applications/sentence-transformers-ft/README.md:
```diff
@@ -0,0 +1,191 @@
+# Sentence Transformers Finetuning
```
will update readme later

justinliang1020 commented on applications/sentence-transformers-ft/modal-optuna.py:
```diff
@@ -0,0 +1,135 @@
+    # TODO: add dropout, seems kinda annoying tho https://github.com/UKPLab/sentence-transformers/issues/677
+
+    # TODO: add learning rate scheduler
+    res = finetune(
```
Might need to modify this to split dataset into "train" "validation" and "test". The model will train on "train" data, Optuna will do grid search using the score of the "validation" dataset, and the final evaluation will be done on "test".

https://stats.stackexchange.com/questions/366862/is-using-both-training-and-test-sets-for-hyperparameter-tuning-overfitting

justinliang1020 committed:
notes

justinliang1020 committed:
rename finetuning

justinliang1020 committed:
add logging to finetuning

justinliang1020 committed:
initial ft embed blog

justinliang1020 committed:
update docs finetune

justinliang1020 committed:
finetune cleanup

justinliang1020 committed:
add base as hyperparameter

justinliang1020 committed:
enable local save hf model

justinliang1020 committed:
editing docs

justinliang1020 commented:
No need to review completely yet, still editing some things but just pushed my changes

## Stuff I did
* added huggingface model compatibility with an example jupyter notebook on using your finetuned model and how to upload to huggingface (will make this more clear in the readme)
* added initial blog post draft (incomplete)
* added logging to finetuning, clarifies that the eval output is the average precision
* added base model as hyperparam for optuna
* did some cleanup on finetune code, renamed filename to Loss function for specificity
* random edits

## Will do
* continue writing initial blog
* run a big optuna hyperparam run to see what happens
* cleanup code
* was thinking about adding wndb

Thoughts for blog: theme should be that fine-tuning methodology is task dependent: i.e. your task determines which dataset format, loss, and evals you use. Also that modal can make fine tuning super easy and that hyperparameter optimization can be used to further train models, especially in the case that base models' MTEB metrics are unreliable (models train on test data for it lol) so we can test multiple base models in parallel.

jxnl commented:
sounds good

justinliang1020 committed:
modal optuna big-ish run

justinliang1020 committed:
finetune docs add modal

justinliang1020 committed:
draft

justinliang1020 committed:
more draft stuff

jxnl commented on applications/sentence-transformers-ft/eval.py:
```diff
@@ -0,0 +1,99 @@
+    "sentence-transformers/all-distilroberta-v1",
+    "sentence-transformers/multi-qa-mpnet-base-dot-v1",
+    "sentence-transformers/multi-qa-distilbert-cos-v1",
+    "llmrails/ember-v1",
```
any jina models?

justinliang1020 replied:
running that rn

justinliang1020 replied:
Getting a weird error with specifically Jina, some remote code error I think which makes it broken on Modal

```
File "/usr/local/lib/python3.11/site-packages/transformers/dynamic_module_utils.py", line 611, in resolve_trust_remote_code
    raise ValueError(
ValueError: The repository for /root/.cache/torch/sentence_transformers/jinaai_jina-embeddings-v2-base-en/ contains custom code which must be executed to correctly load the model. You can inspect the repository content at https://hf.co//root/.cache/torch/sentence_transformers/jinaai_jina-embeddings-v2-base-en/.
Please pass the argument `trust_remote_code=True` to allow custom code to be run.

Stopping app - uncaught exception raised locally: ValueError('The repository for /root/.cache/torch/sentence_transformers/jinaai_jina-embeddings-v2-base-en/ contains custom code which must be executed to correctly load the model. You can inspect the repository content at https://hf.co//root/.cache/torch/sentence_transformers/jinaai_ji
```

jxnl commented:
any tables you can upload?

justinliang1020 commented:
I continued working on the blog post, but I think it should be split into 2 blog posts (basic finetuning and advanced finetuning) since it's covering a lot of content right now.

Part 1 (basic):
* describes the use case for finetuning embedding models
* covers finetuning with Sentence Transformers, specifically covering base model, dataset format, loss, and evaluation
* covers how to use with Modal. specifically how the model can be stored on a modal volume and accessed via Volume CLI. Also what stub and function are (this seems like it has overlap with this not sure if that's an issue https://github.com/567-labs/fastllm/blob/main/docs/embedding.md )

Part 2 (advanced):
* covers what Optuna is, how it can be used for finetuning
* parallel evaluation of base models
* Describes Modal feature (NFS) which can be used to enable distributed hyperparameter finetuning
* what hyperparameters to finetune

Or it could be put into one giant blog post, either way is fine with me.

justinliang1020 commented:
> any tables you can upload?

this contains the data from the `eval.py` run, will format it into readable table for blog

https://github.com/567-labs/fastllm/pull/28/files#diff-ac3eddde1a45fcdc7ffad3dc6c18bab4620173738197e481ccd6a038d0be42fe

Maybe could also try one for the hyperparameter optimization, but will upload that later

justinliang1020 commented on applications/sentence-transformers-ft/eval.py:
```diff
@@ -0,0 +1,99 @@
+
+DATASET_ID = "quora"
+GPU_CONFIG = "a10g"
+MODEL_IDS = [
```
Add our own model to this too, compare it

justinliang1020 committed:
update eval metrics

justinliang1020 committed:
rename files

justinliang1020 committed:
inital README

justinliang1020 committed:
optuna cleanup

justinliang1020 committed:
README update

justinliang1020 commented:
Cleaned up some code, wrote the README, added some example metrics under /examples folder. Also uploaded an example finetuned model here: https://huggingface.co/krunchykat/bge-base-en-v1.5-ft-quora more details in the readme

justinliang1020 committed:
eval metrics readme

jxnl commented:
can you upload to 567-labs 👉 👈

justinliang1020 committed:
change model to 567 labs

justinliang1020 commented:
Changed model to 567 labs https://huggingface.co/567-labs/bge-base-en-v1.5-ft-quora
